### PR TITLE
🍒 [WebAssembly] Emit clangast in custom section aligned by 4 bytes

### PR DESCRIPTION
**Pull request by kateinoigakukun**
Emit __clangast in custom section instead of named data segment
to find it while iterating sections.
This could be avoided if all data segements (the wasm sense) were
represented as their own sections (in the llvm sense).
This can be resolved by https://github.com/WebAssembly/tool-conventions/issues/138

And the on-disk hashtable in clangast needs to be aligned by 4 bytes,
so add paddings in name length field in custom section header.

The length of clangast section name can be represented in 1 byte
by leb128, and possible maximum pads are 3 bytes, so the section
name length won't be invalid in theory.

Fixes https://bugs.llvm.org/show_bug.cgi?id=35928

Differential Revision: https://reviews.llvm.org/D74531

Cherry-pick from https://github.com/llvm/llvm-project/commit/1813fde9cc0b56cee42d9b82e6f22fa00a59cdf9 to build Swift stdlib for WebAssembly importing SwiftShims module.

CC: @MaxDesiatov, @compnerd

### CODE REVIEW

**Comment by kateinoigakukun** [comment]
@swift-ci Please test

**Comment by MaxDesiatov** [comment]
@shahmishal would you be available for review and merge? Thanks!

**Comment by shahmishal** [comment]
cc: @hyp

**Comment by kateinoigakukun** [comment]
Do we need to wait for @hyp 's approval? Or is it ok to merge this for now and revert if it breaks something?
Please let me know if this kind of fix should not be cherry-picked to the swift branch by some policy.

After this patch, swiftwasm can use `apple/llvm-project` instead of its fork and it would improve the maintainability of the project.

**Comment by compnerd** [comment]
Isn't part of this path upstreamed?  I don't think that this should be going here if it is.  I think that waiting for @hyp is probably a good idea here.

**Comment by kateinoigakukun** [comment]
@compnerd This patch has been already landed in the upstream LLVM https://reviews.llvm.org/rG1813fde9cc0b56cee42d9b82e6f22fa00a59cdf9

OK, let's wait @hyp 👍

**Comment by compnerd** [comment]
Ah, I failed to read the truncated link as a cherry-pick.  Thanks!  This is a clean cherry-pick from upstream.  The changes are isolated.  I think that this change is fine to merge, but I'd still prefer to have a cross-repository test first.  Have you run that yet?

**Review by compnerd — APPROVED** [review]

**Comment by kateinoigakukun** [comment]
OK, cross-repos test has been passed now

**Comment by compnerd** [comment]
LGTM, I think that this should be safe to merge.

**Pull request merged**
